bpo-37488 : Document a warning for datetime.utcnow() and utcfromtimestamp()

nanjekyejoannah opened the pull request:
<!-- issue-number: [bpo-37488](https://bugs.python.org/issue37488) -->
https://bugs.python.org/issue37488
<!-- /issue-number -->


Automerge-Triggered-By: @pganssle

pganssle commented on Doc/library/datetime.rst:
```diff
@@ -29,8 +29,9 @@ is used to represent a specific moment in time that is not open to
 A naive object does not contain enough information to unambiguously locate
-itself relative to other date/time objects.  Whether a naive object represents
-Coordinated Universal Time (UTC), local time, or time in some other timezone is
+itself relative to other date/time objects.  Native objects are no longer
```
Was this an intentional change? It looks like you kept in the beginnings of an explanation of the problem, but also the problematic old text.

This is actually a bit complicated to reword, because the sentiment behind it has changed, and we may need to hedge a bit. Let me think about how best to communicate the new intent....

epicfaace commented on Doc/library/datetime.rst:
```diff
@@ -804,6 +811,11 @@ Other constructors, all class methods:
 
+   .. warning::
+
+      Because naive datetimes are treated by many functions as local times.
```
Can you change this wording to be of a similar sentence structure as in the earlier section (for `datetime.utcnow()`)?

epicfaace commented on Misc/NEWS.d/next/Library/2019-09-09-12-48-10.bpo-37488.kteuwA.rst:
```diff
@@ -0,0 +1 @@
+Add warning to datetime.utcnow() and datetime.utcfromtimestamp()
```
```suggestion
Add warning to :meth:`datetime.utcnow` and :meth:`datetime.utcfromtimestamp`
```

aeros commented on Doc/library/datetime.rst:
```diff
@@ -752,6 +752,13 @@ Other constructors, all class methods:
 
+   .. warning::
+
+      Because naive datetimes are treated by many functions as local times,
```
This should be a bit more specific. Also, code comments or Sphinx roles should be used when referencing objects (such as classes, methods, or functions). In this case, code comments would be more appropriate since this is already within the section for ``datetime``. 

```suggestion
      Because naive ``datetime`` objects are treated by many ``datetime`` methods as local times,
```

This line will need to be split off into the next one since it will exceed the reST maximum width of 80 chars (94 chars after the suggestion).

aeros commented on Doc/library/datetime.rst:
```diff
@@ -752,6 +752,13 @@ Other constructors, all class methods:
+   .. warning::
+
+      Because naive datetimes are treated by many functions as local times,
+      it is recommended that you create a timezone-aware datetime
```
Phrasing improvement, it's preferable to avoid the usage of pronouns such as "you" when possible within technical writing, and I think "timezone specific" makes a bit more sense than "timezone-aware". 
```suggestion
      it is recommended to create a timezone specific ``datetime`` object
```

nanjekyejoannah replied:
Thanks for the review. I consulted @pganssle at the sprints and he says we should keep to aware here. I hope I have addressed the other suggestions.

aeros replied:
@nanjekyejoannah:
> Thanks for the review. I consulted @pganssle at the sprints and he says we should keep to aware here. I hope I have addressed the other suggestions.

Sounds good to me. If Paul prefers using "timezone-aware" over "timezone specific", I have no objections.

aeros commented on Doc/library/datetime.rst:
```diff
@@ -804,6 +811,11 @@ Other constructors, all class methods:
+   .. warning::
+
+      Because naive datetimes are treated by many functions as local times.
+      To get an aware :class:`.datetime` object, use :meth:`fromtimestamp`.
```
The Sphinx role here for ``datetime`` can be removed here since this section is already within the ``datetime`` class. The main purpose of the roles is to provide links to other sections, so there's not much of a purpose in a role that links to a section it is already contained within.

```suggestion
      To get a timezone specific datetime object, use :meth:`fromtimestamp`.
```

aeros reviewed (commented):
Thanks for the PR @nanjekyejoannah.

I have a few suggestions:

nanjekyejoannah committed:
Add warning notes in the documentation for utc* methods

This also removes several examples using utc* methods, to avoid
confusion.

bpo-37488: https://bugs.python.org/issue37488

aeros reviewed (approved):
@nanjekyejoannah For consistency, can provide the warning message for each of them?

> Because naive ``datetime`` objects are treated by many ``datetime`` methods
as local times, the recommended way to create aware datetime objects for
the current time in UTC is by calling ``datetime.now(timezone.utc)``.

Otherwise, everything else looks good to me.
